tools/rebase-onto-formatter.sh

```diff
@@ -41,17 +41,56 @@ fi
 
 echo "Running the script now. This might take a while..."
 
-# The idea behind the sscript is ... TODO
+# The general idea is the following:
+#
+# We have a branch consisting of changes C1, C2, ... on top of our BASE_COMMIT 0.
+# Also, from the base commit, a formatting change `f(0)` was made.
+#
+# ```
+#  0 --- C1 ---> 1 --- C2 ---> 2 --- C3 ---> ... --- Cn ---> n
+#  |
+#  f
+#  |
+#  v
+#  0'
+# ```
+#
+# Now, how do we obtain versions of our changes operating on the formatted code (let's call them Ci')?
+#
+# ```
+#  0 --- C1 ---> 1 --- C2 ---> 2 --- C3 ---> ... --- Cn ---> n
+#  |
+#  f
+#  |
+#  v
+#  0' -- C1' --> 1' -- C2' --> 2' -- C3' --> ... --- Cn' --> n'
+# ```
+#
+# One useful thing is that since `f` is defined by an automated tool,
+# we know `f(i)` on every commit `i`.
+#
+# ```
+#  0 --- C1 ---> 1 --- C2 ---> 2 --- C3 ---> ... --- Cn ---> n
+#  |             |             |                             |
+#  f             f             f                             f
+#  |             |             |                             |
+#  v             v             v                             v
+#  0' -- C1' --> 1' -- C2' --> 2' -- C3' --> ... --- Cn' --> n'
+# ```
+#
+# And we can also get its inverse `r(f(i))` by reverting the commit.
+#
+# So we can obtain `Ci' by composing `r(f(i-1)), `Ci` and `f(i)`.
 
 set -x
 
-# edit every commit Ci, adding f(Ci) and r(F(Ci))
+# edit every commit i, adding f(i) and r(f(i))
```

Comment:
I liked `Ci` better!  :)  it's also more consistent with the diagrams above.

Reply:
The problem is that talking about a commit conflates two different ideas:
1. the change/diff it introduces to the previous commit
2. the version of the code at that commit (what you get when you check it out)

In the diagram, these correspond to the arrows and objects respectively. I tried to keep them separate by having the arrows `Ci` and objects `i`.

However, I think in this context this makes it unnecessarily complicated. I'm not trying to define this correctly in terms of category theory (which one could do), but documenting a Bash script. :laughing: 
So I simplified the notation a bit (while trying not to say something too incorrect). I hope it makes it more understandable.